visualizes *cry emote + getting peppersprayed makes you cry

kawoppi opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Using the *cry emote makes tears come out of your eyes for the next 12.8 seconds (4 loops of the animation). You now cry when you get sprayed with pepperspray. [See the demo here.](https://youtu.be/fEAFYNeyZ1M)
![image](https://user-images.githubusercontent.com/94711066/186997862-d88b0788-a44a-447c-9a29-b996e3c2bf93.png)

~~Codewise very similar to [this PR](https://github.com/tgstation/tgstation/pull/66321). I plan on making these not be copy-pastes of eachother in the future, need to figure some stuff out first.~~
The crying overlay is placed by the eyes organ. I went back to the blush overlay and made it so that it's placed by the head bodypart. Also the blush overlay now gets adjusted based on OFFSET_FACE.


<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
It makes the characters more visually expressive. The crying gives pepperspray some more visual feedback, not just when you get hit but while you're recovering from it afterwards as well.
<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: added an animation for using the *cry emote
add: getting peppersprayed causes you to cry
code: the blush overlay is now affected by the face offset
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

kawoppi committed:
added emote visual code very similar to blush visual

kawoppi committed:
updated crying sprite and color

kawoppi committed:
getting peppersprayed makes you cry

kawoppi committed:
updated sprite and color again

kawoppi committed:
changed duration to last 4 animation cycles

kawoppi committed:
removed comment

it's not as important here because the tears sprite doesn't intersect with moth eyes

optimumtact commented:
very soulful change thank you froggers

Timberpoes commented:
Hilarious, I love it.

Kapu1178 commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -602,6 +602,12 @@ GLOBAL_LIST_EMPTY(features_by_species)
 			blush_overlay.color = COLOR_BLUSH_PINK
 			standing += blush_overlay
 
+		// tears
+		if (HAS_TRAIT(species_human, TRAIT_CRYING))
+			var/mutable_appearance/tears_overlay = mutable_appearance('icons/mob/human_face.dmi', "tears", -BODY_ADJ_LAYER)
+			tears_overlay.color = COLOR_DARK_CYAN
+			standing += tears_overlay
```
this should all belong to the head or eyes, it shouldnt be an arbitrary species thing, im trying to get rid of this awful proc.

kawoppi replied:
The crying overlay has been moved to the eyes. I've also added a proc similar to what the eyes have to the head and moved the blush emote to there.

Kapu1178 commented on code/modules/mob/living/carbon/human/emote.dm:
```diff
@@ -1,12 +1,36 @@
+		var/list/key_emotes = GLOB.emote_list["cry"]
+		for(var/datum/emote/living/carbon/human/cry/living_emote in key_emotes)
+			// The existing timer restarts if it's already running
+			visual_timer = addtimer(CALLBACK(living_emote, .proc/end_visual, living_user), CRY_DURATION, TIMER_UNIQUE | TIMER_OVERRIDE)
```
doesnt this need `TIMER_STOPPABLE` to return an ID?

kawoppi replied:
It seems that without that flag it returns the value of TIMER_ID_NULL. With that flag it uses some other number. But it turns out that the timer restarts just fine without needing to track the timer ID. Guess that makes sense because I only assigned it but never used it.

kawoppi committed:
removed timer ID variable

Turns out that this is not needed to restart the existing timer.

kawoppi committed:
tears overlay gets applied by the eyes instead

Can't cry without eyes. They're also affected by face offsets now.

kawoppi committed:
blush overlay gets applied by the head instead

kawoppi committed:
eye offsets are now changed in a foreach loop

Not doing it for the head because some things in there may need to use the head offset instead of the face offset.

Kapu1178 commented on code/modules/surgery/bodyparts/head.dm:
```diff
@@ -180,6 +180,31 @@
+#define OFFSET_X 1
+#define OFFSET_Y 2
+
+/// Generates a list of overlays that the head should display for the given parent
+/obj/item/bodypart/head/proc/generate_body_overlay(mob/living/carbon/human/parent)
+	if(!istype(parent) || parent.get_bodypart(BODY_ZONE_HEAD) != src)
+		CRASH("Generating a body overlay for [src] targeting an invalid parent '[parent]'.")
+
+	var/list/overlays = list()
+
+	// Blush emote overlay
+	if (HAS_TRAIT(parent, TRAIT_BLUSHING)) // Caused by either the *blush emote or the "drunk" mood event
+		var/mutable_appearance/blush_overlay = mutable_appearance('icons/mob/human_face.dmi', "blush", -BODY_ADJ_LAYER) // Should appear behind the eyes
+		blush_overlay.color = COLOR_BLUSH_PINK
+		overlays += blush_overlay
+		if(OFFSET_FACE in parent.dna?.species.offset_features)
+			var/offset = parent.dna.species.offset_features[OFFSET_FACE]
+			blush_overlay.pixel_x += offset[OFFSET_X]
+			blush_overlay.pixel_y += offset[OFFSET_Y]
+
+	return overlays
+
+#undef OFFSET_X
+#undef OFFSET_Y
```
This already exists. its called `get_limb_icon()` (and its sister proc which obtains render data, `update_limb()`). Limbs should not require any DNA information from the parent after they're generated.

The general pipeline for a bodypart is that when it's created & attached, `update_limb(is_creating = TRUE)` is called. This caches all of the information required to draw it, which `get_limb_icon()` does, which is called by `carbon/proc/update_body_parts()`

Kapu1178 commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -561,6 +561,9 @@ GLOBAL_LIST_EMPTY(features_by_species)
 	var/obj/item/bodypart/head/noggin = species_human.get_bodypart(BODY_ZONE_HEAD)
 
 	if(noggin && !(HAS_TRAIT(species_human, TRAIT_HUSK)))
+		for(var/head_overlay in noggin.generate_body_overlay(species_human))
+			standing += head_overlay
+
```
This is somehow worse, since you now have a superfluous 2-way relationship between the head and a species datum, which is bad.

Palpetinus commented:
>The crying overlay is placed by the eyes organ. 

Does this mean if the mob has no eyes, it doesn't have the visual either?
What about robotic eyes?

kawoppi commented:
> Does this mean if the mob has no eyes, it doesn't have the visual either? What about robotic eyes?

There's no crying animation if your eyes have been removed, it doesn't forbid you from doing the emote though. The crying animation still works with robotic eyes

kawoppi committed:
moved blush visual to the proper procs

kawoppi committed:
forgot to put that newline back

Kapu1178 commented on code/modules/surgery/bodyparts/head.dm:
```diff
@@ -182,35 +184,32 @@
 #undef OFFSET_X
 #undef OFFSET_Y
 
+	if (is_blushing != HAS_TRAIT(owner, TRAIT_BLUSHING)) // Caused by either the *blush emote or the "drunk" mood event
+		is_blushing = HAS_TRAIT(owner, TRAIT_BLUSHING)
+		owner.update_body_parts()
+
```
Do not call `update_body_parts()` in `update_limb()`. Call it where it needs to be updated (like the blush emote itself).

Kapu1178 reviewed (changes requested):
I meant to hit request changes not approve

kawoppi committed:
blushing uses update_body_parts()

kawoppi committed:
restricted crying animation to humans as well

Kapu1178 reviewed (approved):
Looks good I think.

CapybaraExtravagante commented:
whats what the awfully specific duration O.o

kawoppi commented:
> whats what the awfully specific duration O.o

Oh right I probably should've put that in the description. The animation has a total delay of 32, that makes for 3.2 seconds.
![image](https://user-images.githubusercontent.com/94711066/187518783-28d7cce8-1fa0-40db-8f98-22bbf2e59c9b.png)
Looping the animation 3 times makes for 9.6 seconds. The last of the pepperspray effects wear off in 10 seconds so one more loop is added and that brings it to 12.8.

RaveRadbury commented:
Does it work on top of clown and mime masks?

kawoppi commented:
> Does it work on top of clown and mime masks?

I don't think I can make that work right now without causing bugs unfortunately. I can make it appear on top if you cry while wearing a clown/mime mask. But then you can swap out your mask for gas mask and it would appear on top as well. And if you put the clown/mime mask on after crying it will still appear under the mask. If I can manage to make a component for these emote visuals then I think I could make it work properly but that will have to happen later.
